File: parallel-qsort/PQsort.erl
-module(pqsort).
-export([pqsort/1]).

pqsort_worker(Listener, []) -> Listener ! {self(), []};
pqsort_worker(Listener, [H | T]) ->
    PID_1 = spawn_link(pqsort, fun pqsort_worker/2, [self(), [ X || X <- T, H >= X ]]),
    PID_2 = spawn_link(pqsort, fun pqsort_worker/2, [self(), [ Y || Y <- T, H < Y ]]),
    receive
        {PID_1, List_1} ->
            receive
                {PID_2, List_2} -> Listener ! {self(), List_1 ++ [H] ++ List_2}
            end;
        {PID_2, List_2} ->
            receive
                {PID_1, List_1} -> Listener ! {self(), List_1 ++ [H] ++ List_2}
            end
    end.

pqsort(List) -> element(2, pqsort_worker(self(), List)).

File: parallel-qsort/PQsort.ex
defmodule PQsort do
    def pqsort(parent, []), do: send(parent, {self(), []})
    def pqsort(parent, [head | tail]) do
        pid_one = spawn_link(PQsort, :pqsort, [self(), (for x <- tail, head >= x, do: x)])
        pid_two = spawn_link(PQsort, :pqsort, [self(), (for y <- tail, head < y, do: y)])
        receive do
            {^pid_one, small_list} ->
                receive do
                    {^pid_two, large_list} -> send(parent, {self(), small_list ++ [head] ++ large_list})
                end
            {^pid_two, large_list} ->
                receive do
                    {^pid_one, small_list} -> send(parent, {self(), small_list ++ [head] ++ large_list})
                end
        end
    end
    def pqsort(list), do: elem(pqsort(self(), list), 1)
end